**Natural Language Process (NLP)**

**Natural Language** -> Türkçe, İngilizce gibi insanların konuştuğu dillerdir.

**Natural Language Process** -> Bilgisayar ve insan arasındaki etkileşimi sağlar. Machine learning modelleri textlere uygulanır.

Bir text mining işlemidir.

Tweetler, ansiklopediler, kitaplarda bulunan textlerden anlam çıkarma işlemidir.

------------------------------------------------------------------------------------------

**cleaning data** 

"*merHAba__" bu ifadenin merhaba olduğunu makinenin anlaması için cleaning işlemi yapılır.

--------------------------------------------------------------------------------------------

**bag of words**

Feature çıkartma yöntemlerinden bir tanesidir. Yani textin machine learning uygulanabilecek hale getirilmesidir.

--------------------------------------------------------------------------------------------

**text classification**

Bir metnin dilinin belirlenmesi, atılan tweetlerin göre cinsiyet analizi yapılması text classification örneğidir.

--------------------------------------------------------------------------------------------

NLP kullanım alanlarından bazıları, 

*   **prediction genre of book** : kitapların türünü predict ederken kullanılır. 
*   **question/answering** : siride olduğu gibi konuşmanın anlaşılıp cevap verilmesidir.
*   **document summarization** : bir dokümanın, bir kitabın özetinin çıkarılma işlemidir.

--------------------------------------------------------------------------------------------

**NLP kütüphaneleri**,

nltk

spacy

standford nlp

open nlp






In [ ]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv(r"/content/drive/MyDrive/data/gender_classifier.csv", encoding = "latin1")
#encoding = "latin1" datada latin harflerinin olduğu belirtilip, bu dikkate alarak dataframe dönüştürülür. 
#r parametresi read demektir.

Datada bulunan description feature'ı kullanılarak, bu description'ı yapanın kadın mı erkek mi olduğu belirlenecektir.

Description kısmında text dosyaları bulunmaktadır.

In [ ]:
data = pd.concat([data.gender, data.description],axis=1) 

Bütün datayı kullanmak yerine datada kullanılacak sütunlar ayrılır.

In [ ]:
data.head()

gender                                        description
0    male                              i sing my own rhythm.
1    male  I'm the author of novels filled with family dr...
2    male                louis whining and squealing and all
3    male  Mobile guy.  49ers, Shazam, Google, Kleiner Pe...
4  female  Ricky Wilson The Best FRONTMAN/Kaiser Chiefs T...

In [ ]:
data.isnull().sum()

gender           97
description    3744
dtype: int64

Datada missing valuelar ve bazı anlamsız tweetler bulunmaktadır.

In [ ]:
data.dropna(axis=0, inplace=True)
#datada bulunan missing valuelar temizlenir.
#missing valuelar satırda olduğundan, axis=0 ile satır olarak çıkartılır.
#inplace=True ile bu temizleme işlemi kalıcı hale getirilir.

Genderlar female ve male olarak string şekilde bulunmaktadır.

Classification yapılması için integer ya da boolean tipinde olmalıdır.

In [ ]:
data.gender = [1 if each=="female" else 0 for each in data.gender]
#for each döngüsüyle female 1'e, male 0'a çevrilir.

# **Regular Expression (RE)**

In [ ]:
import re 
first_description = data.description[4]

In [ ]:
first_description

'Ricky Wilson The Best FRONTMAN/Kaiser Chiefs The Best BAND Xxxx Thank you Kaiser Chiefs for an incredible year of gigs and memories to cherish always :) Xxxxxxx'

Cleaning kısmı datadaki 4.satır üzerinden gösterilip, tüm dataya uygulanacaktır.

In [ ]:
description = re.sub("[^a-zA-z]"," ",first_description)
#a-zA-z bu kısımda küçük a'dan küçük z'ye ve büyük a'dan büyük z'ye kadar olan harfler bulunur.
#'^' başına bu işaret geldiğinde bulma anlamına gelir.
#harfler dışındaki karakterlerin yerine ( ':' ve ')' gibi) boşluk getirilir.
#bu işlemler sub methoduyla yapılır.

regular expression ile bir search pattern bulunur.

küçük a'dan küçük z'ye ve büyük a'dan büyük z'ye kadar olan harflerin dışındaki karakterler re parametresiyle bulunmuştur.

Buradaki search pattern '^a-zA-z' bu kısımdır.

Regular experession ile textin içinde farklı aramalar yapılır.

In [ ]:
description

'Ricky Wilson The Best FRONTMAN Kaiser Chiefs The Best BAND Xxxx Thank you Kaiser Chiefs for an incredible year of gigs and memories to cherish always    Xxxxxxx'

Text'e bakıldığında harf olmayan karakterlerin ':' ve ')' kaldırıldığı ve bunların yerine boşluk getirildiği görülür.

Bilgisayar dili büyük harfle yazılan kelimeler ile küçük harfle yazılan kelimeleri farklı kelime olarak algılar.

In [ ]:
description = description.lower()
#lower methoduyla tüm harfler küçük harfe çevrilir.

In [ ]:
description

'ricky wilson the best frontman kaiser chiefs the best band xxxx thank you kaiser chiefs for an incredible year of gigs and memories to cherish always    xxxxxxx'

Tüm harflerin küçük harfe çevrildiği görülür.

# **Irrelavant Words (Stopwords)**

Makinenin textten anlam çıkarırken 'the', 'and', 'or', 'as' gibi kelimeler önem arz etmemektedir.

Mesela burada yapılacak classificationda 'and', 'the' gibi kelimeler texti yazanın kadın mı erkek mi olduğuna dair bir yardımda bulunmamaktadır. Bu kelimeleri kadın da yazabilir, erkek de yazabilir.

Bu kelimeler textten çıkartılarak filtreleme işlemi yapılır.

In [ ]:
import nltk   #natural language tool kit

nltk.download("stopwords") #stop wordler downloand edilir.
                           #stopwordler bilgisayarda corpus adında bir klasöre indirilir.
from nltk.corpus import stopwords #stopwordler corpus klasöründen import edilir.

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Textin içindeki her bir kelime bir listenin içerisinde depolanıp stop wordler ile karşılaştırıldıktan sonra eşleşme olursa bu kelimeler atılacaktır. 

In [ ]:
description = description.split()
#split methodunun default değeri textteki kelimeleri bir listenin içerisinde teker teker ayırır.

In [ ]:
description

['ricky',
 'wilson',
 'the',
 'best',
 'frontman',
 'kaiser',
 'chiefs',
 'the',
 'best',
 'band',
 'xxxx',
 'thank',
 'you',
 'kaiser',
 'chiefs',
 'for',
 'an',
 'incredible',
 'year',
 'of',
 'gigs',
 'and',
 'memories',
 'to',
 'cherish',
 'always',
 'xxxxxxx']

Split methodu yerine tokenize methodu da kullanılabilir.

In [ ]:
nltk.download('punkt')
description = nltk.word_tokenize(description)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Split yerine tokenize methodu tercih edilir. Çünkü textte couldn't gibi kelimeler olduğunda split methodunda 'couldn't' diye ayrılırken tokenize methodunda 'could', 'not' diye ayrılır.

Çünkü split methodu boşluklara göre ayırdığından bu ayrıma varamaz.

In [ ]:
description = [ word for word in description if not word in set(stopwords.words("english"))]
#set methoduyla stopwordlerin unique'leri bulunur.
#list comprehension ile textteki kelimelerde gezinip eğer stop word varsa textten atılır.

In [ ]:
description

['ricky',
 'wilson',
 'best',
 'frontman',
 'kaiser',
 'chiefs',
 'best',
 'band',
 'xxxx',
 'thank',
 'kaiser',
 'chiefs',
 'incredible',
 'year',
 'gigs',
 'memories',
 'cherish',
 'always',
 'xxxxxxx']

Textten 'the', 'of', 'an' gibi kelimelerin atıldığı görülür.

# **Lemmatization**

Bir textteki 'okul', 'okulu', 'okula' kelimeleri makine için 3 farklı kelimeyi ifade etmektedir.

Bu durum text classificaton yapılmasını zorlaştırır ve feature sayısını gereksiz artırır.

Makineye bu kelimelerin tek bir 'okul' kelimesi olarak öğretilmesi gerekir.

Bunun için kelimeler köklerine ayrılır.(loved -> love, gitmeyeceğim -> git)

In [ ]:
from nltk.stem import WordNetLemmatizer
lemma = WordNetLemmatizer()
description = [lemma.lemmatize(word) for word in description]

In [ ]:
description

['ricky',
 'wilson',
 'best',
 'frontman',
 'kaiser',
 'chief',
 'best',
 'band',
 'xxxx',
 'thank',
 'kaiser',
 'chief',
 'incredible',
 'year',
 'gig',
 'memory',
 'cherish',
 'always',
 'xxxxxxx']

Kelimelerin köklerine ayrıldığı görülür.

memories -> memory

chiefs -> chief

gigs -> gig

In [ ]:
description = " ".join(description)

Boşluk stringiyle kelimeler birleştirilir.

In [ ]:
description

'ricky wilson best frontman kaiser chief best band xxxx thank kaiser chief incredible year gig memory cherish always xxxxxxx'

Buraya kadar yapılan adımlarda;
*   datadaki gereksiz karakterler kaldırıldı.
*   tüm harfler küçük harfe çevrildi.
*   stop wordler kaldırıldı
*   Kelimelerin kökleri bulundu.







Bu işlemler tek bir cümleye uygulanmıştır. 

Bütün texte uygulanacaktır.

In [ ]:
description_list = []

for description in data.description:
    description = re.sub("[^a-zA-Z]", " ", description)  
    #Gereksiz karakterlerin yerine boşluk koyulur.
    
    description = description.lower()  #Tüm harfler küçük harfe çevrilir.

    description = nltk.word_tokenize(description)  #text kelimelere ayrılır.
    
    description = [ word for word in description if not word in set(stopwords.words("english"))]
    #stop wordler listeden atılarak filtreleme yapılır.

    lemma = WordNetLemmatizer()
    description = [lemma.lemmatize(word) for word in description] #kelimelerin kökü bulunur.
    
    description = " ".join(description)
    #kelimelerin arasına boşluk koyularak tekrardan birleştirilir. Böylelikle text elde edilmiş olur.
    
    description_list.append(description)

# **Bag of Words**

Üç tane cümle olsun.

1) Merhaba ben okula geldim

2) Merhaba sen okula git

3) Merhaba ben okula döndüm

İlk cümledeki her bir kelime bir feature'dır.

İkinci cümlede 'Merhaba' ve 'okula' kelimeleri bir önceki cümlede vardır.

Bu kelimeler tekrardan feature olarak alınmaz, kalan kelimeler alınır.

Bu işlem tüm cümlelere uygulanır.



In [ ]:
data.shape

(16224, 2)

Datada 16224 satır bulunmaktadır. Bu da 16224 cümle demektir. Her birinde farklı kelimeler olsa hemen hemen 32000 kelime demektir. Burada 32000 kelimenin içerisinden en çok kullanılan 500 tanesi alınacaktır. Çünkü diğer durumda algoritma çok zorlanıp, yavaşlayacaktır.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer 
# CountVectorizer bag of words yaratmak için kullanılan methoddur.
max_features = 500 #sınır belirlenir.
count_vectorizer = CountVectorizer(max_features = max_features, stop_words = "english") 
#İngilizce'deki stop wordler cümlelerin içinden çıkartılır.
sparce_matrix = count_vectorizer.fit_transform(description_list).toarray()
#count_vectorizer modele uyarlanıp, sparce_matrix'e eşitlenir.

print("En sık kullanilan {} kelimeler: {}".format(max_features,count_vectorizer.get_feature_names()))

En sık kullanilan 500 kelimeler: ['account', 'activist', 'actor', 'addict', 'adult', 'adventure', 'advocate', 'alum', 'amazing', 'america', 'american', 'angel', 'animal', 'anime', 'app', 'area', 'art', 'artist', 'ask', 'aspiring', 'author', 'award', 'away', 'awesome', 'baby', 'bad', 'band', 'based', 'beautiful', 'beauty', 'beer', 'believe', 'best', 'better', 'big', 'bio', 'bit', 'bitch', 'black', 'blog', 'blogger', 'blue', 'book', 'booking', 'born', 'bot', 'boy', 'brand', 'breaking', 'building', 'business', 'ca', 'car', 'care', 'cat', 'cause', 'ceo', 'certified', 'change', 'channel', 'check', 'chicago', 'chief', 'child', 'christ', 'christian', 'city', 'class', 'club', 'coach', 'coffee', 'college', 'com', 'come', 'comic', 'coming', 'communication', 'community', 'company', 'computer', 'conservative', 'consultant', 'contact', 'content', 'continuous', 'control', 'cool', 'country', 'county', 'crazy', 'create', 'creative', 'creator', 'culture', 'currently', 'dad', 'daily', 'dance', 'data', '

sparce matrix 1'lerden ve 0'lardan oluşur.

Yukarıdaki cümlelerdeki unique kelimeler yan yana yazılır.

'Merhaba ben okula geldim sen git döndüm'

Birinci cümle için ilk 4 kelime birinci cümlede geçtiğinden 1 yazılır, diğer kelimeler geçmediğinden 0 yazılır.

Bu işlem diğer cümleler için de aynıdır.

# **Text Classification**

Şu anda dataset kelimelerden oluşmaktadır.

In [ ]:
y = data.iloc[:,0].values #cinsiyet classı numpy array'e çevrilir.
x = sparce_matrix

In [ ]:
#train test split
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.1,random_state=42)

Naive Bayes algoritması kullanılacaktır.

In [ ]:
#model fit
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb.fit(x_train,y_train)

GaussianNB()

In [ ]:
#prediction
y_pred = nb.predict(x_test)

In [ ]:
#accuracy
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy of naive bayes classification: {}".format(accuracy))

Accuracy of naive bayes classification: 0.585335797905114


Data 500 kelime kullanılarak train edilmişti. 

Burada kullanılan feature sayısıyla %58'lik bir doğruluk elde edilmiştir.